### ISSUE TITLE
add configuration property to support custom components with no lowercase transformation

### ISSUE THREAD

**Issue by frandevel** [cure53/DOMPurify#739]
> If you're working with React JS and you're allowing external HTML to be rendered, you need to use the `dangerouselySetInnerHtml` property. If this is the case and you want to support custom components in this HTML piece, you will have something like `<Icon name="home"></Icon>` in the HTML, and will need the JSXParser library to recognize this component and render it accordingly. Passing this HTML through the DOMPurify sanitization means that `<Icon name="home"></Icon>` will be transformed into `<icon name="home"></icon>`, which the JSXParser won't have registered and, therefore, won't be rendered.

### Background & Context

JSX Parser library: https://www.npmjs.com/package/react-jsx-parser

### Bug

Custom "HTML" tags names being transformed into lowercase makes DOMPurify incompatible with rendering custom React components.
If I use the PARSER_MEDIA_TYPE: 'application/xhtml+xml', I get this preserved, but all links (<a href....) stripped out from the HTML, what is not desirable, and also the name space being inserted in all components as a new attribute:
`<Icon name="home" xmlns=\"http://www.w3.org/1999/xhtml\"></Icon>"`

#### Input

`<Icon name="home"></Icon>`

#### Given output

`<icon name="home"></icon>`

#### Expected output

`<Icon name="home"></Icon>`

### Feature

It would be very helpful to be able to have some configuration property like "PRESERVE_CASE", to avoid having this transformation. 
As an example:
```
const dirty = `<div>
                            <Icon name="home" size="xl">Some text</Icon>
                            <ProgressWheel progress="75"></ProgressWheel>
                        </div>`
const SUPPORTED_CUSTOM_COMPONENTS = ['Icon', 'ProgressWheel'];
const SUPPORTED_CUSTOM_ATTRIBUTES = ['size', 'progress']
const sanitizationConfig = {
			ADD_TAGS: SUPPORTED_CUSTOM_COMPONENTS,
			ADD_ATTR: SUPPORTED_CUSTOM_ATTRIBUTES,
                        **PRESERVE_CASE: SUPPORTED_CUSTOM_COMPONENTS**
		};
const clean = DOMPurify.sanitize(dirty, sanitizationConfig);
```

**Comment by frandevel** [comment]
@cure53 I can try, we're currently a bit overloaded with our platform release, but I will try ASAP. Thanks for the quick answer

**Comment by cure53** [comment]
Haha, same here, it's Q4 after all :D However, I had a look. I think this can be done quite easily.

```javascript
    // HTML tags and attributes are not case-sensitive, converting to lowercase. Keeping XHTML as is.
    transformCaseFunc =
      PARSER_MEDIA_TYPE === 'application/xhtml+xml'
        ? stringToString
        : stringToLowerCase;
```

This should be the place to get started at.

**Comment by frandevel** [comment]
Great to read this @cure53 , thanks for having a look. 
Actually, I would expect that we don't even need to use that `PARSER_MEDIA_TYPE`. I guess we should try to stick to `HTML` parsing and just preserve the casing for the custom components that might be specified in a string array. 
Or am I missing anything?

**Comment by cure53** [comment]
Yes, you are absolutely right. It would just make sense to add a condition here. If either the parser type is `'application/xhtml+xml'` or if the config says `PRESERVE_CASE` then we resort to `stringToString` and leave the `tagName` as is.

**Comment by frandevel** [comment]
Ah sorry. You were talking about a quick workaround now. Let me test that. Thanks!

**Comment by cure53** [comment]
I am currently running tests with a boolean `PRESERVE_CASE` and it looks good, if you wanna pick that up and implement the rest, I'd not say no :D

**Comment by cure53** [comment]
Hmmm, this is trickier than I thought. The problem is that, if we use default `PARSER_MEDIA_TYPE`, then DOMParser will transform everything to lowercase.

If we use `PARSER_MEDIA_TYPE: 'application/xhtml+xml'` then you will get namespace attributes.

**Comment by frandevel** [comment]
I see. What about using the `CUSTOM_ELEMENT_HANDLING`? Maybe we could have a specific `preserveCase` property in the `tagNameCheck` object and behave accordingly. 
Also we must consider that using the `PARSER_MEDIA_TYPE: 'application/xhtml+xml'` will result in all `<a href...>` elements being stripped out in the sanitization.

**Comment by cure53** [comment]
Custom elements, by spec, need to contain a dash in the tagName, no?

**Comment by cure53** [comment]
I think, given that DOMPurify utilizes the browser and React decided to go its own path, we might not have an easy way to get this to work. Unless I am overlooking something right now.

**Comment by frandevel** [comment]
`PRESERVE_CASE: ['Icon', 'ProgressWheel']`
I guess we need to use `stringToString` in case of using `PARSER_MEDIA_TYPE === 'application/xhtml+xml'` OR in case the element name is in the `PRESERVE_CASE` array of strings

**Comment by cure53** [comment]
Aye, I just fear that DOMParser might mess that up :D But, if you get it to work, I'd be more than happy to review the PR and merge in case all is fine :)

**Comment by cure53** [comment]
Closing this now as there is no easy solution, please reopen if needed or any ideas came up :)

**Comment by ShahriarKh** [comment]
This is indeed required. Previously, I used `renderToString` and `JsxParser`. What I used to do was to do a regex search on the raw data, find Jsx elements, give the found jsx element to jsx parser, convert the parsed jsx to string by using `renderToString` and then sanitize the whole parsed content:
```jsx
function convertJsxToHtml(content) {
   // find every jsx element and return renderToString(<JsxParser jsx={foundElement}/>)
}

<div dangerouslySetInnerHtml={{ __html: sanitize(convertJsxToHtml(content)) }} >
```

However, I gave up because renderToString doesn't return script tags correctly and gives hydration error and a bunch of other headaches.
